### PR TITLE
Add retry behavior to GitHub client

### PR DESCRIPTION
**Pull request by alzabo**
The problem this solves, and it may not be a problem that others are having, is that sometimes lint-review can't connect to my GitHub Enterprise instance. I suspect that the problem is on the GitHub side, but another team administers it, etc.

The symptom I see is that some PRs get stuck indefinitely with a lint-review "Pending" status. Logs on my instance look like the following:

```
worker_1  | [2018-10-12 16:34:05,219: ERROR/Worker-1] ('Connection aborted.', BadStatusLine("''",))
worker_1  | Traceback (most recent call last):
worker_1  |   File "/code/lintreview/tasks.py", line 55, in process_pull_request
worker_1  |     processor.load_changes()
worker_1  |   File "/code/lintreview/processor.py", line 32, in load_changes
worker_1  |     files = self._pull_request.files()
worker_1  |   File "/code/lintreview/repo.py", line 130, in files
worker_1  |     return list(self.pull.files())
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/github3/structs.py", line 67, in __iter__
worker_1  |     headers=headers)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/github3/models.py", line 185, in _get
worker_1  |     return self.session.get(url, **kwargs)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 525, in get
worker_1  |     return self.request('GET', url, **kwargs)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/github3/session.py", line 88, in request
worker_1  |     response = super(GitHubSession, self).request(*args, **kwargs)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 512, in request
worker_1  |     resp = self.send(prep, **send_kwargs)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 622, in send
worker_1  |     r = adapter.send(request, **kwargs)
worker_1  |   File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 495, in send
worker_1  |     raise ConnectionError(err, request=request)
worker_1  | ConnectionError: ('Connection aborted.', BadStatusLine("''",))
```

I don't do a very good job of tracking metrics on how many pull requests lint-review tests for us, but I normally see this type of failure happen a couple times a week.

### CODE REVIEW

**Comment by alzabo** [comment]
I've got this patch running in my environment, where it seems to be working fine.

Obviously, there are some magic numbers and stuff that maybe should be configurable. I figured I'd get the PR open and get your input.

**Comment by markstory** [comment]
Looks like a bunch of the test fixtures/mocks will also need to be updated as Github3.py has changed their implementation around a bit.

**Comment by alzabo** [comment]
lol, I was a little surprised that so many of the tests were failing, but admittedly, I fudged running many of the tests locally because my Python install is semi-busted.

I'll try to find some time this week to fix what I can figure out. It looks like the classes from tests want a session object passed into `__init__`, which is easy enough to do.

**Comment by alzabo** [comment]
@markstory I got the tests working again, and also updated the constructor for the Retry operator to be configurable.

File: tests/__init__.py [hunk 1]
```diff
@@ -19,11 +20,13 @@ def load_fixture(filename):
 
 
 def create_pull_files(data):
-    return [PullFile(f) for f in json.loads(data)]
+    session = GitHubSession()
+    return [PullFile(f, session) for f in json.loads(data)]
```
Comment (markstory) [inline tests/__init__.py]: Are these constructor changes part of the github3.py upgrade?

Comment (alzabo) [reply tests/__init__.py]: Yeah, the individual classes like `PullFile` all descend from [github3.models.GitHubCore](https://github.com/sigmavirus24/github3.py/blob/162f247b4664a2612b2cfd959497678b069f2c5e/src/github3/models.py#L29), which expects a session to be passed into `__init__`.

I'm not sure how important it is to have a valid session object in each test case, e.g. PullFile doesn't seem to have any methods that require a valid session, but certain functionality, like constructing URLs, seems to be delegated to the object's session.

An example that has many calls to [GitHubCore._build_url](https://github.com/sigmavirus24/github3.py/blob/162f247b4664a2612b2cfd959497678b069f2c5e/src/github3/models.py#L224-L226) is the [_Repository class](https://github.com/sigmavirus24/github3.py/blob/162f247b4664a2612b2cfd959497678b069f2c5e/src/github3/repos/repo.py#L49), which is the base class for a couple different repository classes.

Comment (alzabo) [reply tests/__init__.py]: Also, I tried to account for the various changes to tests and some of the "why" in the commit message for https://github.com/markstory/lint-review/pull/233/commits/161d81b27f6e297d654e2714bb245ea834415d91 . I know it's a big patch and some of the changes like the updated json files aren't super easy to review by looking at a diff.

**Comment by markstory** [comment]
:clap: Thank you

**Pull request merged**